server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java

```diff
@@ -120,6 +107,13 @@ public void tearDown() throws Exception {
 
     public void testSnapshotShardSizes() throws Exception {
         final int maxConcurrentFetches = randomIntBetween(1, 10);
+
+        final AtomicInteger rerouteCount = new AtomicInteger(0);
```

Comment:
I wonder, couldn't we get rid of the busy asserting completely by doing this via a latch (has the slight upside that we verify exact count a little stronger with the assertion on the latch count?)?

e.g.:

```diff
diff --git a/server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java b/server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java
index a4483bd5ce2..fdcdcdb013f 100644
--- a/server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/snapshots/InternalSnapshotsInfoServiceTests.java
@@ -120,12 +120,20 @@ public class InternalSnapshotsInfoServiceTests extends ESTestCase {
 
     public void testSnapshotShardSizes() throws Exception {
         final int maxConcurrentFetches = randomIntBetween(1, 10);
+
+        final int numberOfShards = randomIntBetween(1, 50);
+        final CountDownLatch rerouteLatch = new CountDownLatch(numberOfShards);
+        final RerouteService rerouteService = (reason, priority, listener) -> {
+            listener.onResponse(clusterService.state());
+            assertThat(rerouteLatch.getCount(), greaterThanOrEqualTo(0L));
+            rerouteLatch.countDown();
+        };
+
         final InternalSnapshotsInfoService snapshotsInfoService =
             new InternalSnapshotsInfoService(Settings.builder()
                 .put(INTERNAL_SNAPSHOT_INFO_MAX_CONCURRENT_FETCHES_SETTING.getKey(), maxConcurrentFetches)
                 .build(), clusterService, () -> repositoriesService, () -> rerouteService);
 
-        final int numberOfShards = randomIntBetween(1, 50);
         final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
         final long[] expectedShardSizes = new long[numberOfShards];
         for (int i = 0; i < expectedShardSizes.length; i++) {
@@ -163,12 +171,10 @@ public class InternalSnapshotsInfoServiceTests extends ESTestCase {
 
         latch.countDown();
 
-        assertBusy(() -> {
-            assertThat(snapshotsInfoService.numberOfKnownSnapshotShardSizes(), equalTo(numberOfShards));
-            assertThat(snapshotsInfoService.numberOfUnknownSnapshotShardSizes(), equalTo(0));
-            assertThat(snapshotsInfoService.numberOfFailedSnapshotShardSizes(), equalTo(0));
-        });
-        verify(rerouteService, times(numberOfShards)).reroute(anyString(), any(Priority.class), any());
+        assertTrue(rerouteLatch.await(10L, TimeUnit.SECONDS));
+        assertThat(snapshotsInfoService.numberOfKnownSnapshotShardSizes(), equalTo(numberOfShards));
+        assertThat(snapshotsInfoService.numberOfUnknownSnapshotShardSizes(), equalTo(0));
+        assertThat(snapshotsInfoService.numberOfFailedSnapshotShardSizes(), equalTo(0));
         assertThat(getShardSnapshotStatusCount.get(), equalTo(numberOfShards));
 
         final SnapshotShardSizeInfo snapshotShardSizeInfo = snapshotsInfoService.snapshotShardSizes();
```

Just a suggestion :) this approach looks fine as well.

Reply:
This is a good suggestion, thanks!